File: leetcode/algorithms/clone_graph/clone_graph.cc
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (node == NULL) {
            return node;
        }

        map<UndirectedGraphNode*, UndirectedGraphNode*> clones;

        queue<UndirectedGraphNode*> q;
        q.push(node);

        UndirectedGraphNode* head = new UndirectedGraphNode(node->label);
        clones[node] = head;

        while (!q.empty()) {
            UndirectedGraphNode* top = q.front();
            q.pop();

            for (int i = 0; i < top->neighbors.size(); ++i) {
                UndirectedGraphNode* other = top->neighbors[i];
                if (clones.count(other) == 0) {
                    UndirectedGraphNode* new_other = new UndirectedGraphNode(other->label);
                    clones[top]->neighbors.push_back(&new_other);
                    clones[other] = new_other;
                    q.push(other);
                } else {
                    clones[top]->neighbors->push_back(clones[other]);
                }
            }
        }

        return head;
    }
};

File: leetcode/algorithms/clone_graph/clone_graph.py
import collections

class UndirectedGraphNode(object):
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    def cloneGraph(self, node):
        if node is None:
            return node

        self.clones = {}

        queue = collections.deque()
        queue.append(node)

        new_node = UndirectedGraphNode(node.label)
        self.clones[node] = new_node

        while queue:
            top = queue.popleft()

            for other in top.neighbors:
                other_copy = self.clones.get(other, None) or UndirectedGraphNode(other.label)
                self.clones[top].neighbors.append(other_copy)

                if not other in self.clones:
                    self.clones[other] = other_copy
                    queue.append(other)

        return new_node
